### ISSUE TITLE
Commands could take (lazily evaluated) expressions

### ISSUE THREAD

**Issue by andychu** [oilshell/oil#955]
https://oilshell.zulipchat.com/#narrow/stream/121540-oil-discuss/topic/Syntax.20idea.20for.20structured.20pipelines

```
myls | where [size > 10]
```

on your wiki: `Oil Pipeline Syntax`

**Comment by andychu** [comment]
Also this might be an opportunity to make proc signatures use more names  and fewer punctuatino characters

```
# optional Str

proc p(src, dest Str, myexpr Expr, @rest, myblock Block) {
  echo $src $dest
}
```

In expression mode, we could also have

```
const myarglist = &[a, b, c]
```

**Comment by andychu** [comment]
Philosophically, this is a little like Eggex.  Bringing more syntax into the shell proper, rather than having every tool invent its own syntax

**Comment by andychu** [comment]
note that this conflicts with `[ -z "" ]`, so it will probably have to be in `oil:all`.  I think that's fine; people can set `shopt --set parse_brackets` interactively if they want.  It could be the LAST thing in their oshrc.

**Comment by andychu** [comment]
Problems this solves

- select/filter syntax for QTT (data frames)
- Awk dialect lazy evaluation: `when [x > 10] { echo "big" }`
- json read/write  syntax: `json write [d]` and `json write [{name: 'foo'}]`

**Comment by andychu** [comment]
Two issues with brackets:

- it can be confused with JSON arrays
  - e.g. `json dump [[1,2,3]]` is ugly
  - https://lobste.rs/s/wprseq/on_unix_composability#c_qdosqu
- it can be confused with type parameters, which is ugly
  - `schema { field foo [List[Int]] }`

I think it's better to use parens, and probably possible as long as they are **procs** and not **funcs**:


```
json dump ([1,2,3])

schema Foo {
  field bar (List[Int])
}

```

Also:

```
wok {
  when (x > 10) {
    echo $x
  }
```

Yeah I guess this is possible since you never need procs with zero arguments and parens!  That is, `foo()` is never needed; it's just `foo`.

**Comment by andychu** [comment]
Note: I tried this, and I remember the lookahead to distinguish it from  a shell style function was hard

```
f () { echo hi; }  # need to look ahead TWICE

f (x == y)
```

Also this is a different code path:

```
f arg (x == y)
```

We *could* use `f .(x == y)`, although it's inconsistent because the thing in the "sigil pair" is in *expression mode* and not *command mode*, in contrast to `$(mycmd)`, `@(mycmd)`, and `^(mycmd)`

**Comment by andychu** [comment]
Possible solution for the lookahead problem: Make `'(' whitespace? ')'`  a single token with re2c!  `Id.Op_ParenPair`

This isn't valid anywhere else in command mode, e.g. `foo bar ()` is invalid

Then we just look ahead one token, and handle the rest of the command uniformly.  Yes I think this should work.

**Comment by andychu** [comment]
Problem with this idea: it messes up parsing arrays!    Like `a=()`

Hm.  I suppose we could special case that too

**Comment by andychu** [comment]
- More precise lookahead is done
- Fixed the `$identity (x)` bug (space was allowed)

But we still have the potential issue of

```
$identity(x)
```

vs.

```
$identity (x)
```

which now becomes valid.  It's a different in the space.

One idea is to create a MANDATORY space for lazily evaluated expressions.  Yeah I think that is acceptable.

We could possibly document it with "warts"

The shell language and Oil language is space sensitive in a few other places ...

**Comment by andychu** [comment]
This is in good shape, next step is #1024

**Comment by andychu** [comment]
Released with Oil 0.9.5 - http://www.oilshell.org/blog/2021/11/recent-progress.html
